types/underscore/underscore-tests.ts

```diff
@@ -165,372 +165,423 @@ declare const level2UnionList: _.List<_.List<string | number>>;
+_.sortBy(numberArray, num => Math.sin(num)); // $ExpectType number[]
 
 // grouping numbers by their non-fractional parts
-_([1.3, 2.1, 2.4]).groupBy((e) => Math.floor(e));
+_([1.3, 2.1, 2.4]).groupBy((e) => Math.floor(e)); // $ExpectType Dictionary<number[]>
 
 // grouping numbers by the value of a specified property
-_.groupBy(['one', 'two', 'three'], 'length');
+_.groupBy(['one', 'two', 'three'], 'length'); // $ExpectType Dictionary<string[]>
 
 // indexing items in a dictionary by age
-_.indexBy(stooges, 'age');
+_.indexBy(stooges, 'age'); // $ExpectType Dictionary<{ name: string; age: number; }>
 
 // counting numbers by their evenness
-_.countBy([1, 2, 3, 4, 5], (num) => (num % 2 == 0) ? 'even' : 'odd');
+_.countBy(numberArray, num => (num % 2 === 0) ? 'even' : 'odd'); // $ExpectType Dictionary<number>
 
 // shuffling numbers
-_.shuffle([1, 2, 3, 4, 5, 6]);
+_.shuffle(numberArray); // $ExpectType number[]
 
 // converting an array-like structure to an array
-(function (a, b, c, d) { return _.toArray(arguments).slice(1); })(1, 2, 3, 4);
+(function (a, b, c, d) {
+    const args = _.toArray(arguments); // $ExpectType any[]
+    return args;
+})(1, 2, 3, 4);
 
 // determining the number of items in a dictionary
-_.size({ one: 1, two: 2, three: 3 });
+_.size(explicitNumberDictionary); // $ExpectType number
 
 // splitting numbers into sets of even and odd values
-_.partition([0, 1, 2, 3, 4, 5], (num) => {return num % 2 === 0 });
-
-interface Family {
-    name: string;
-    relation: string;
-}
+_.partition(numberArray, num => num % 2 === 0); // $ExpectType [number[], number[]]
 
 // creating a function that can determine if one object's property values matches another's
-var isUncleMoe = _.matches<Family>({ name: 'moe', relation: 'uncle' });
-var uncleMoe: Family = { name: 'moe', relation: 'uncle' };
-isUncleMoe(uncleMoe);
+{
+    const isUncleMoe = _.matches({ name: 'moe', relation: 'uncle' }); // $ExpectType Predicate<{ name: string; relation: string; }>
+    isUncleMoe({ name: 'moe', relation: 'uncle' }); // $ExpectType boolean
+}
 
 ///////////////////////////////////////////////////////////////////////////////////////
 
 // retrieving the first item in an array
-_.first([5, 4, 3, 2, 1]);
+_.first(numberArray); // $ExpectType number | undefined
 
 // retrieving all but the last element in an array
-_.initial([5, 4, 3, 2, 1]);
+_.initial(numberArray); // $ExpectType number[]
 
 // retrieving the last two elements in an array
-_.last([5, 4, 3, 2, 1], 2);
+_.last(numberArray, 2); // $ExpectType number[]
 
 // retrieving all but the first two elements in an array
-_.rest([5, 4, 3, 2, 1], 2);
+_.rest(numberArray, 2); // $ExpectType number[]
 
 // removing falsy values
-_.compact([0, 1, false, 2, '', 3, undefined]);
+_.compact([0, 1, false, 2, '', '3', undefined]); // $ExpectType (string | number | true)[]
 
 // deep flattening an array
-_.flatten([1, [2], [3, [[4]]]]);
+_.flatten([[[1, 2], [3]], [[4, 5]]]); // $ExpectType any[]
```

Comment:
This is a friendlier example for shallow flattening since it results in just `number[][]` instead of `(number | number[][])[]` like the original example did, and I felt that it was nice to stay consistent with the example between shallow and deep flattening to show how they differ.